# PyTITAN 
## Tool for In-depth Transcription factor binding affinity ANalysis

Python hybridized with a bit of R for analyzing gain/loss of binding of transcription factors using a list of genomic variants as input.

The procedure is to some extent based on the approach in pjshort's denovoTF package (https://github.com/pjshort/denovoTF). 

For further information on that package and its use see:
> Short PJ, McRae JF, Gallone G, Sifrim A, Won H, Geschwind DH, Wright CF, Firth HV, Fitzpatrick DR, Barrett JC, Hurles ME. 2018. De novo mutations in regulatory elements in neurodevelopmental disorders. _Nature_. Volume 555. Pages 611-616. https://doi.org/10.1038/nature25983

# Dependencies

PyTITAN was developed for Unix environments. We suggest running python >= 3.7 and R >= 3.6 on an Ubuntu >= 16.04 machine (or similar).

## Python packages

- pandas >= 0.25.3
- pyfaidx >= 0.5.6
- numpy >= 1.17.4

## R packages

- TFBSTools >= 1.24.0
- JASPAR2020 >= 0.99.8

Note: you can also use different versions of the JASPAR database (tested: JASPAR2018 1.1.1, JASPAR2016 1.14.0, JASPAR2014 1.22.0).

## Further required data

PyTITAN requires a reference genome in FASTA format. For example, the GRCh37 (hg19) reference genome can be found here
> ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/reference/phase2_reference_assembly_sequence/hs37d5.fa.gz

# Installation

Download or clone this repository.

Run
```
Rscript extract_matrix.R pfm
```
To obtain the position frequency matrices for all transcription factors. 

If you wish to use your own data, make sure that it follows the expected file naming format: [TF-ID]\_[TF-Name]\_[TF-strand (+/-)].pfm.

# How to

## Input

PyTITAN expects input in the following format (tab-separated):
| CHROM | POS | REF | ALT |
| --- | --- | --- | --- |
| 12 | 66358347 | C | T |

An additional column 'ID' (containing a _unique_ ID) is optional but will be generated by PyTITAN if it is missing.

## Using the tool

Run
```
python3 main.py \
            -i variant_file.tsv \
            -f pwm_folder \
            -o output_file.tsv \
            -g genome \
            -t int \ # optional
            -m float \ # optional
            -n float \ # optional
            -p int \ # optional
            -q float \ # optional
            -b float float float float # optional
```

## Options

| Parameter | Description | Value | Default |
| --- | --- | --- | --- |
| -i | Path to input file | String | - |
| -f | Path to folder containing the PFM data created by the extract_matrix.R script | String | - |
| -o | Path to output file | String | - |
| -g | Path to reference genome file (FASTA) | String | - |
| -t | Threshold (percentile) for binding affinity of a variant position / transcription factor binding site combination to be included. | Integer | 50 |
| -m | Minimum information content of position | Float | 1.0 |
| -n | Minimum information content of base (either reference or alternative) | Float | 0.5 |
| -p | Number of processes to use | Integer | 1 |
| -q | Pseudocount for PWM generation | Float | 0.8 |
| -b | Nucleotide background probabilities: A C G T | Float Float Float Float | 0.25 each |

For further information on the command line, run
```
python3 main.py -h
```

# PWM creation

The creation of PWMs follows the principles described in:
> Wassermann WW, Sandelin A. 2004. Applied bioinformatics for the identification of regulatory elements. _Nat Rev Genet_. Volume 5. Pages 276-287. https://doi.org/10.1038/nrg1315

By default, a pseudocount value of 0.8 is used, as suggested here:
> Nishida K, Frith MC, Nakai K. 2008. Pseudocounts for transcription factor binding sites. _Nucleic Acids Res_. Volume 37. Issue 3. Pages 939-944. https://doi.org/10.1093/nar/gkn1019

By default, the following values are used for the nucleotide background probabilities:

| A | C | G | T |
| --- | --- | --- | --- |
| 0.25 | 0.25 | 0.25 | 0.25 |
